src/docker.c

```diff
@@ -0,0 +1,1137 @@
+  for (int i = 0; i < stats->num_cpus; i++) {
+    value_t used, idle, used_percent, idle_percent;
+    unsigned int cpu_count = i + 1;	// 1 indexed
+    char *cpu_num = (char *) calloc (11, sizeof(char));
+    ssnprintf(cpu_num, 10, "%d", cpu_count);
+    used.counter = stats->percpu_usage[i];
+    idle.counter = stats->percpu_idle[i];
+    used_percent.gauge = stats->percpu_percent_used[i];
+    idle_percent.gauge = stats->percpu_percent_idle[i];
+    dispatch_value_list(hostname, "cpu", cpu_num, "cpu", "used", NULL, 1, used);
+    dispatch_value_list(hostname, "cpu", cpu_num, "cpu", "idle", NULL, 1, idle);
+    dispatch_value_list(hostname, "cpu", cpu_num, "percent", "used", NULL, 1,
+        used_percent);
+    dispatch_value_list(hostname, "cpu", cpu_num, "percent", "idle", NULL, 1,
+        idle_percent);
+  }
+}
+
+static void dispatch_container_memory_stats(memory_stats_t * stats,
+    char *hostname) {
+  unsigned long bytes_results[] = {
+    stats->usage,
+    stats->free,
+  };
+
+  unsigned long percent_results[] = {
+    stats->used_percentage,
+    stats->free_percentage,
+  };
+
+  assert(STATIC_ARRAY_SIZE(bytes_results) == STATIC_ARRAY_SIZE(percent_results));
+  for (int i = 0; i < STATIC_ARRAY_SIZE(bytes_results); i++) {
+    value_t bytes, percent;
+    bytes.gauge = bytes_results[i];
+    percent.gauge = percent_results[i];
+    dispatch_value_list(hostname, "memory", NULL, "memory",
+        MEMORY_METRIC_TYPES[i], NULL, 1, bytes);
+    dispatch_value_list(hostname, "memory", NULL, "percent",
+        MEMORY_METRIC_TYPES[i], NULL, 1, percent);
+  }
+}
+
+static void dispatch_container_interface_stats(interface_stats_t * stats,
+    char *hostname) {
+  unsigned long rx_results[] = {
+    stats->rx_bytes,
+    stats->rx_packets,
+    stats->rx_errors,
+  };
+
+  unsigned long tx_results[] = {
+    stats->tx_bytes,
+    stats->tx_packets,
+    stats->tx_errors,
+  };
+
+  const char *types[] = {
+    "if_octets",
+    "if_packets",
+    "if_errors",
+  };
+
+  assert(STATIC_ARRAY_SIZE(rx_results) == STATIC_ARRAY_SIZE(INTERFACE_KEYS));
+  assert(STATIC_ARRAY_SIZE(tx_results) == STATIC_ARRAY_SIZE(INTERFACE_KEYS));
+  for (int i = 0; i < STATIC_ARRAY_SIZE(INTERFACE_KEYS); i++) {
+    value_t values[2];
+    values[0].derive = rx_results[i];
+    values[1].derive = tx_results[i];
+    dispatch_value_list(hostname, "interface", stats->name, types[i], NULL, NULL,
+                        2, values[0], values[1]);
+  }
+}
+
+static void dispatch_container_network_stats(network_stats_t * stats,
+    char *hostname) {
+  for (int i = 0; i < stats->count; i++) {
+    if (stats->interfaces[i] != NULL) {
+      dispatch_container_interface_stats(stats->interfaces[i], hostname);
+    }
+  }
+}
+
+static void dispatch_container_stats(container_resource_t * resource) {
+  char *hostname = (char *) calloc (76, sizeof(char));
+  ssnprintf(hostname, 75, "container.%s", resource->id);
```

Comment:
This might be worth a comment somewhere (or maybe in many places) to let folks know that hostname is expected to contain the container (resource?) id if we are reporting stats for a container.